From e0a102e5f0dd7a6d8ab5b8a1efcca54bca87d25a Mon Sep 17 00:00:00 2001
From: mssx86 <mss@waifu.club>
Date: Thu, 1 Aug 2019 05:08:27 +0300
Subject: disable archive compression patch.

---
 deblob-main | 32 --------------------------------
 1 file changed, 32 deletions(-)

diff --git a/deblob-main b/deblob-main
index 80c8ba9..db41324 100644
--- a/deblob-main
+++ b/deblob-main
@@ -265,42 +265,10 @@ xargs tar --append -f linux-libre-$kver-$gnu.tar
 echo Wiping out extracted trees
 rm -rf linux-$kver orig-linux-$kver
 
-echo Creating vcdiff between linux-$kver.tar and linux-libre-$kver-$gnu.tar
-xdelta3 -e -9 -S djw -s linux-$kver.tar linux-libre-$kver-$gnu.tar linux-libre-$kver-$gnu.vcdiff || : # don't fail if xdelta3 is not present
-
-echo Creating xdelta between linux-$kver.tar and linux-libre-$kver-$gnu.tar
-xdelta delta -0 linux-$kver.tar linux-libre-$kver-$gnu.tar linux-libre-$kver-$gnu.xdelta || : # xdelta returns nonzero on success
-
 cleanup="linux-libre-$kver-$gnu.tar linux-libre-$kver-$gnu.xdelta"
 
 echo Compressing binary deltas and linux-libre-$kver-$gnu.tar
 rm -f linux-$kver.tar
-if test -f linux-libre-$kver-$gnu.xdelta; then
-  bzip2 -k9 linux-libre-$kver-$gnu.xdelta
-  xz -k9 linux-libre-$kver-$gnu.xdelta || :
-  lzip -k9s64MiB linux-libre-$kver-$gnu.xdelta || :
-fi
-bzip2 -k9 linux-libre-$kver-$gnu.tar
-xz -k9 linux-libre-$kver-$gnu.tar || :
-lzip -k9s64MiB linux-libre-$kver-$gnu.tar || :
-
-echo Done except for signing, feel free to interrupt
-for f in \
-  linux-libre-$kver-$gnu.tar \
-  linux-libre-$kver-$gnu.tar.bz2 \
-  linux-libre-$kver-$gnu.tar.xz \
-  linux-libre-$kver-$gnu.tar.lz \
-  linux-libre-$kver-$gnu.vcdiff \
-  linux-libre-$kver-$gnu.xdelta \
-  linux-libre-$kver-$gnu.xdelta.bz2 \
-  linux-libre-$kver-$gnu.xdelta.xz \
-  linux-libre-$kver-$gnu.xdelta.lz \
-; do
-  if test -f $f; then
-    gpg -a --detach-sign $f
-    mv $f.asc $f.sign
-  fi
-done
 
 rm -f $cleanup
 cleanup=
-- 
2.20.1

