diff -uraN a/platform/js/src/old-configure.in b/platform/js/src/old-configure.in
--- a/platform/js/src/old-configure.in	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/js/src/old-configure.in	2020-08-09 07:52:27.092459574 +0300
@@ -1935,11 +1935,6 @@
 dnl ========================================================
 AC_MSG_CHECKING([for -pipe support])
 if test -n "$GNU_CC" -a -n "$GNU_CXX"; then
-    dnl Any gcc that supports firefox supports -pipe.
-    CFLAGS="$CFLAGS -pipe"
-    CXXFLAGS="$CXXFLAGS -pipe"
-    AC_MSG_RESULT([yes])
-else
     AC_MSG_RESULT([no])
 fi
 
diff -uraN a/platform/media/webrtc/trunk/build/common.gypi b/platform/media/webrtc/trunk/build/common.gypi
--- a/platform/media/webrtc/trunk/build/common.gypi	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/media/webrtc/trunk/build/common.gypi	2020-08-09 07:46:02.634553150 +0300
@@ -2197,7 +2197,6 @@
           # Don't export any symbols (for example, to plugins we dlopen()).
           # Note: this is *required* to make some plugins work.
           '-fvisibility=hidden',
-          '-pipe',
         ],
         'cflags_cc': [
           '-fno-rtti',
diff -uraN a/platform/media/webrtc/trunk/peerconnection_client.target.mk b/platform/media/webrtc/trunk/peerconnection_client.target.mk
--- a/platform/media/webrtc/trunk/peerconnection_client.target.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/media/webrtc/trunk/peerconnection_client.target.mk	2020-08-09 07:43:32.461808484 +0300
@@ -56,7 +56,6 @@
 	-Wno-unused-parameter \
 	-Wno-missing-field-initializers \
 	-fvisibility=hidden \
-	-pipe \
 	-fPIC \
 	-pthread \
 	-I/usr/include/gtk-2.0 \
@@ -146,7 +145,6 @@
 	-Wno-unused-parameter \
 	-Wno-missing-field-initializers \
 	-fvisibility=hidden \
-	-pipe \
 	-fPIC \
 	-pthread \
 	-I/usr/include/gtk-2.0 \
diff -uraN a/platform/media/webrtc/trunk/testing/gmock.target.mk b/platform/media/webrtc/trunk/testing/gmock.target.mk
--- a/platform/media/webrtc/trunk/testing/gmock.target.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/media/webrtc/trunk/testing/gmock.target.mk	2020-08-09 07:40:22.668867352 +0300
@@ -31,7 +31,6 @@
 	-Wno-unused-parameter \
 	-Wno-missing-field-initializers \
 	-fvisibility=hidden \
-	-pipe \
 	-fPIC \
 	-Wno-format \
 	-Wno-unused-result \
@@ -80,7 +79,6 @@
 	-Wno-unused-parameter \
 	-Wno-missing-field-initializers \
 	-fvisibility=hidden \
-	-pipe \
 	-fPIC \
 	-Wno-format \
 	-Wno-unused-result \
diff -uraN a/platform/media/webrtc/trunk/testing/gmock_main.target.mk b/platform/media/webrtc/trunk/testing/gmock_main.target.mk
--- a/platform/media/webrtc/trunk/testing/gmock_main.target.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/media/webrtc/trunk/testing/gmock_main.target.mk	2020-08-09 07:40:38.932948001 +0300
@@ -31,7 +31,6 @@
 	-Wno-unused-parameter \
 	-Wno-missing-field-initializers \
 	-fvisibility=hidden \
-	-pipe \
 	-fPIC \
 	-Wno-format \
 	-Wno-unused-result \
@@ -79,7 +78,6 @@
 	-Wno-unused-parameter \
 	-Wno-missing-field-initializers \
 	-fvisibility=hidden \
-	-pipe \
 	-fPIC \
 	-Wno-format \
 	-Wno-unused-result \
diff -uraN a/platform/media/webrtc/trunk/testing/gtest.target.mk b/platform/media/webrtc/trunk/testing/gtest.target.mk
--- a/platform/media/webrtc/trunk/testing/gtest.target.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/media/webrtc/trunk/testing/gtest.target.mk	2020-08-09 07:40:51.417009906 +0300
@@ -30,7 +30,6 @@
 	-Wno-unused-parameter \
 	-Wno-missing-field-initializers \
 	-fvisibility=hidden \
-	-pipe \
 	-fPIC \
 	-Wno-format \
 	-Wno-unused-result \
@@ -77,7 +76,6 @@
 	-Wno-unused-parameter \
 	-Wno-missing-field-initializers \
 	-fvisibility=hidden \
-	-pipe \
 	-fPIC \
 	-Wno-format \
 	-Wno-unused-result \
diff -uraN a/platform/media/webrtc/trunk/testing/gtest_main.target.mk b/platform/media/webrtc/trunk/testing/gtest_main.target.mk
--- a/platform/media/webrtc/trunk/testing/gtest_main.target.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/media/webrtc/trunk/testing/gtest_main.target.mk	2020-08-09 07:41:02.713065920 +0300
@@ -31,7 +31,6 @@
 	-Wno-unused-parameter \
 	-Wno-missing-field-initializers \
 	-fvisibility=hidden \
-	-pipe \
 	-fPIC \
 	-Wno-format \
 	-Wno-unused-result \
@@ -78,7 +77,6 @@
 	-Wno-unused-parameter \
 	-Wno-missing-field-initializers \
 	-fvisibility=hidden \
-	-pipe \
 	-fPIC \
 	-Wno-format \
 	-Wno-unused-result \
diff -uraN a/platform/media/webrtc/trunk/testing/gtest_prod.host.mk b/platform/media/webrtc/trunk/testing/gtest_prod.host.mk
--- a/platform/media/webrtc/trunk/testing/gtest_prod.host.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/media/webrtc/trunk/testing/gtest_prod.host.mk	2020-08-09 07:39:56.768738920 +0300
@@ -29,7 +29,6 @@
 	-Wno-unused-parameter \
 	-Wno-missing-field-initializers \
 	-fvisibility=hidden \
-	-pipe \
 	-fPIC \
 	-Wno-format \
 	-Wno-unused-result \
@@ -74,7 +73,6 @@
 	-Wno-unused-parameter \
 	-Wno-missing-field-initializers \
 	-fvisibility=hidden \
-	-pipe \
 	-fPIC \
 	-Wno-format \
 	-Wno-unused-result \
diff -uraN a/platform/media/webrtc/trunk/testing/gtest_prod.target.mk b/platform/media/webrtc/trunk/testing/gtest_prod.target.mk
--- a/platform/media/webrtc/trunk/testing/gtest_prod.target.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/media/webrtc/trunk/testing/gtest_prod.target.mk	2020-08-09 07:40:10.916809076 +0300
@@ -29,7 +29,6 @@
 	-Wno-unused-parameter \
 	-Wno-missing-field-initializers \
 	-fvisibility=hidden \
-	-pipe \
 	-fPIC \
 	-Wno-format \
 	-Wno-unused-result \
@@ -74,7 +73,6 @@
 	-Wno-unused-parameter \
 	-Wno-missing-field-initializers \
 	-fvisibility=hidden \
-	-pipe \
 	-fPIC \
 	-Wno-format \
 	-Wno-unused-result \
diff -uraN a/platform/nsprpub/configure b/platform/nsprpub/configure
--- a/platform/nsprpub/configure	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/nsprpub/configure	2020-08-09 07:55:44.457438254 +0300
@@ -5332,48 +5332,6 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking for gcc -pipe support" >&5
 $as_echo_n "checking for gcc -pipe support... " >&6; }
 if test -n "$GNU_CC" && test -n "$GNU_CXX" && test -n "$GNU_AS"; then
-    echo '#include <stdio.h>' > dummy-hello.c
-    echo 'int main() { printf("Hello World\n"); return 0; }' >> dummy-hello.c
-    ${CC} -S dummy-hello.c -o dummy-hello.s 2>&5
-    cat dummy-hello.s | ${AS} -o dummy-hello.S - 2>&5
-    if test $? = 0; then
-        _res_as_stdin="yes"
-    else
-        _res_as_stdin="no"
-    fi
-    if test "$_res_as_stdin" = "yes"; then
-        _SAVE_CFLAGS=$CFLAGS
-        CFLAGS="$CFLAGS -pipe"
-        cat confdefs.h - <<_ACEOF >conftest.$ac_ext
-/* end confdefs.h.  */
- #include <stdio.h>
-int
-main ()
-{
-printf("Hello World\n");
-  ;
-  return 0;
-}
-_ACEOF
-if ac_fn_c_try_compile "$LINENO"; then :
-  _res_gcc_pipe="yes"
-else
-  _res_gcc_pipe="no"
-fi
-rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
-        CFLAGS=$_SAVE_CFLAGS
-    fi
-    if test "$_res_as_stdin" = "yes" && test "$_res_gcc_pipe" = "yes"; then
-        _res="yes";
-        CFLAGS="$CFLAGS -pipe"
-        CXXFLAGS="$CXXFLAGS -pipe"
-    else
-        _res="no"
-    fi
-    rm -f dummy-hello.c dummy-hello.s dummy-hello.S dummy-hello a.out
-    { $as_echo "$as_me:${as_lineno-$LINENO}: result: $_res" >&5
-$as_echo "$_res" >&6; }
-else
     { $as_echo "$as_me:${as_lineno-$LINENO}: result: no" >&5
 $as_echo "no" >&6; }
 fi
diff -uraN a/platform/nsprpub/configure.in b/platform/nsprpub/configure.in
--- a/platform/nsprpub/configure.in	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/nsprpub/configure.in	2020-08-09 07:54:06.536952693 +0300
@@ -642,34 +642,6 @@
 dnl ========================================================
 AC_MSG_CHECKING([for gcc -pipe support])
 if test -n "$GNU_CC" && test -n "$GNU_CXX" && test -n "$GNU_AS"; then
-    echo '#include <stdio.h>' > dummy-hello.c
-    echo 'int main() { printf("Hello World\n"); return 0; }' >> dummy-hello.c
-    ${CC} -S dummy-hello.c -o dummy-hello.s 2>&5
-    cat dummy-hello.s | ${AS} -o dummy-hello.S - 2>&5
-    if test $? = 0; then
-        _res_as_stdin="yes"
-    else
-        _res_as_stdin="no"
-    fi
-    if test "$_res_as_stdin" = "yes"; then
-        _SAVE_CFLAGS=$CFLAGS
-        CFLAGS="$CFLAGS -pipe"
-        AC_TRY_COMPILE( [ #include <stdio.h> ],
-            [printf("Hello World\n");],
-            [_res_gcc_pipe="yes"],
-            [_res_gcc_pipe="no"] )
-        CFLAGS=$_SAVE_CFLAGS
-    fi
-    if test "$_res_as_stdin" = "yes" && test "$_res_gcc_pipe" = "yes"; then
-        _res="yes";
-        CFLAGS="$CFLAGS -pipe"
-        CXXFLAGS="$CXXFLAGS -pipe"
-    else
-        _res="no"
-    fi
-    rm -f dummy-hello.c dummy-hello.s dummy-hello.S dummy-hello a.out
-    AC_MSG_RESULT([$_res])
-else
     AC_MSG_RESULT([no])
 fi
 
diff -uraN a/platform/old-configure.in b/platform/old-configure.in
--- a/platform/old-configure.in	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/old-configure.in	2020-08-09 07:54:22.217030446 +0300
@@ -4444,11 +4444,6 @@
 dnl ========================================================
 AC_MSG_CHECKING([for -pipe support])
 if test -n "$GNU_CC" -a -n "$GNU_CXX"; then
-    dnl Any gcc that supports firefox supports -pipe.
-    CFLAGS="$CFLAGS -pipe"
-    CXXFLAGS="$CXXFLAGS -pipe"
-    AC_MSG_RESULT([yes])
-else
     AC_MSG_RESULT([no])
 fi
 
diff -uraN a/platform/security/nss/coreconf/BeOS.mk b/platform/security/nss/coreconf/BeOS.mk
--- a/platform/security/nss/coreconf/BeOS.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/security/nss/coreconf/BeOS.mk	2020-08-09 07:41:20.609154662 +0300
@@ -32,7 +32,7 @@
 	OPTIMIZER	= -O2
 endif
 
-OS_CFLAGS		= $(DSO_CFLAGS) $(OS_REL_CFLAGS) -Wall -Wno-switch -pipe
+OS_CFLAGS		= $(DSO_CFLAGS) $(OS_REL_CFLAGS) -Wall -Wno-switch
 OS_LIBS			= -lbe
 
 DEFINES			+= -DBEOS
diff -uraN a/platform/security/nss/coreconf/Darwin.mk b/platform/security/nss/coreconf/Darwin.mk
--- a/platform/security/nss/coreconf/Darwin.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/security/nss/coreconf/Darwin.mk	2020-08-09 07:43:10.465699411 +0300
@@ -85,7 +85,7 @@
 # definitions so that the linker can catch multiply-defined symbols.
 # Also, common symbols are not allowed with Darwin dynamic libraries.
 
-OS_CFLAGS	= $(DSO_CFLAGS) $(OS_REL_CFLAGS) -fno-common -pipe -DDARWIN -DHAVE_STRERROR -DHAVE_BSD_FLOCK $(DARWIN_SDK_CFLAGS)
+OS_CFLAGS	= $(DSO_CFLAGS) $(OS_REL_CFLAGS) -fno-common -DDARWIN -DHAVE_STRERROR -DHAVE_BSD_FLOCK $(DARWIN_SDK_CFLAGS)
 
 ifdef BUILD_OPT
 ifeq (11,$(ALLOW_OPT_CODE_SIZE)$(OPT_CODE_SIZE))
diff -uraN a/platform/security/nss/coreconf/Linux.mk b/platform/security/nss/coreconf/Linux.mk
--- a/platform/security/nss/coreconf/Linux.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/security/nss/coreconf/Linux.mk	2020-08-09 07:43:19.617744794 +0300
@@ -130,7 +130,7 @@
 OS_PTHREAD = -lpthread 
 endif
 
-OS_CFLAGS		= $(DSO_CFLAGS) $(OS_REL_CFLAGS) $(ARCHFLAG) -pipe -ffunction-sections -fdata-sections -DHAVE_STRERROR
+OS_CFLAGS		= $(DSO_CFLAGS) $(OS_REL_CFLAGS) $(ARCHFLAG) -ffunction-sections -fdata-sections -DHAVE_STRERROR
 ifeq ($(KERNEL),Linux)
 	OS_CFLAGS	+= -DLINUX -Dlinux
 endif
diff -uraN a/platform/security/nss/coreconf/NCR3.0.mk b/platform/security/nss/coreconf/NCR3.0.mk
--- a/platform/security/nss/coreconf/NCR3.0.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/security/nss/coreconf/NCR3.0.mk	2020-08-09 07:42:00.237351168 +0300
@@ -16,7 +16,7 @@
 ###
 
 RANLIB           = true
-GCC_FLAGS_EXTRA += -pipe
+GCC_FLAGS_EXTRA += 
 
 DEFINES		+= -DSVR4 -DSYSV -DHAVE_STRERROR -DNCR
 
diff -uraN a/platform/security/nss/coreconf/NetBSD.mk b/platform/security/nss/coreconf/NetBSD.mk
--- a/platform/security/nss/coreconf/NetBSD.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/security/nss/coreconf/NetBSD.mk	2020-08-09 07:42:13.337416127 +0300
@@ -26,7 +26,7 @@
 DLL_SUFFIX		= so.1.0
 endif
 
-OS_CFLAGS		= $(DSO_CFLAGS) $(OS_REL_CFLAGS) -Wall -Wno-switch -pipe -DNETBSD -Dunix -DHAVE_STRERROR -DHAVE_BSD_FLOCK
+OS_CFLAGS		= $(DSO_CFLAGS) $(OS_REL_CFLAGS) -Wall -Wno-switch -DNETBSD -Dunix -DHAVE_STRERROR -DHAVE_BSD_FLOCK
 
 OS_LIBS			= -lcompat
 
diff -uraN a/platform/security/nss/coreconf/OpenBSD.mk b/platform/security/nss/coreconf/OpenBSD.mk
--- a/platform/security/nss/coreconf/OpenBSD.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/security/nss/coreconf/OpenBSD.mk	2020-08-09 07:41:44.341272343 +0300
@@ -26,7 +26,7 @@
 
 DLL_SUFFIX		= so.1.0
 
-OS_CFLAGS		= $(DSO_CFLAGS) $(OS_REL_CFLAGS) -Wall -Wno-switch -pipe -DOPENBSD
+OS_CFLAGS		= $(DSO_CFLAGS) $(OS_REL_CFLAGS) -Wall -Wno-switch -DOPENBSD
 
 OS_LIBS			= 
 
diff -uraN a/platform/security/nss/coreconf/QNX.mk b/platform/security/nss/coreconf/QNX.mk
--- a/platform/security/nss/coreconf/QNX.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/security/nss/coreconf/QNX.mk	2020-08-09 07:41:34.261222359 +0300
@@ -27,7 +27,7 @@
 	OPTIMIZER	= -O2
 endif
 
-OS_CFLAGS		= $(DSO_CFLAGS) $(OS_REL_CFLAGS) -Vgcc_ntox86 -Wall -pipe -DNTO -DHAVE_STRERROR -D_QNX_SOURCE -D_POSIX_C_SOURCE=199506 -D_XOPEN_SOURCE=500
+OS_CFLAGS		= $(DSO_CFLAGS) $(OS_REL_CFLAGS) -Vgcc_ntox86 -Wall -DNTO -DHAVE_STRERROR -D_QNX_SOURCE -D_POSIX_C_SOURCE=199506 -D_XOPEN_SOURCE=500
 
 ifdef USE_PTHREADS
 	DEFINES		+= -D_REENTRANT
diff -uraN a/platform/security/nss/coreconf/ReliantUNIX.mk b/platform/security/nss/coreconf/ReliantUNIX.mk
--- a/platform/security/nss/coreconf/ReliantUNIX.mk	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/security/nss/coreconf/ReliantUNIX.mk	2020-08-09 07:42:23.305465556 +0300
@@ -14,7 +14,7 @@
 	AS          = $(CC)
 	ASFLAGS     += -x assembler-with-cpp
 	LD          = gld
-	ODD_CFLAGS  = -pipe -Wall -Wno-format -Wno-switch
+	ODD_CFLAGS  = -Wall -Wno-format -Wno-switch
 	ifdef BUILD_OPT
 		OPTIMIZER += -O6
 	endif
diff -uraN a/platform/security/nss/coreconf/config.gypi b/platform/security/nss/coreconf/config.gypi
--- a/platform/security/nss/coreconf/config.gypi	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/security/nss/coreconf/config.gypi	2020-08-09 07:42:55.853626954 +0300
@@ -346,7 +346,6 @@
           'OTHER_CFLAGS': [
             '-fPIC',
             '-fno-common',
-            '-pipe',
           ],
         },
         'conditions': [
@@ -390,7 +389,6 @@
           [ 'OS!="mac" and OS!="ios" and OS!="win"', {
             'cflags': [
               '-fPIC',
-              '-pipe',
               '-ffunction-sections',
               '-fdata-sections',
             ],
diff -uraN a/platform/security/nss/tests/bogo/bogo.sh b/platform/security/nss/tests/bogo/bogo.sh
--- a/platform/security/nss/tests/bogo/bogo.sh	2020-07-30 12:37:13.000000000 +0300
+++ b/platform/security/nss/tests/bogo/bogo.sh	2020-08-09 07:42:38.589541346 +0300
@@ -44,7 +44,7 @@
 SOURCE_DIR="$(cd "$cwd"/../..; pwd -P)"
 bogo_init
 (cd "$BORING"/ssl/test/runner;
- GOPATH="$cwd" go test -pipe -shim-path "${BINDIR}"/nss_bogo_shim \
+ GOPATH="$cwd" go test -shim-path "${BINDIR}"/nss_bogo_shim \
 	 -loose-errors -allow-unimplemented \
 	 -shim-config "${SOURCE_DIR}/gtests/nss_bogo_shim/config.json") \
 	 2>bogo.errors | tee bogo.log
