From 5c4a8fd90c9e23e9f72fdbb0185e1e67e81bf9eb Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Tue, 4 Apr 2023 19:05:01 +0300
Subject: [PATCH 02/15] libcxx fixes.

---
 memory/mozalloc/throw_gcc.h | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/memory/mozalloc/throw_gcc.h b/memory/mozalloc/throw_gcc.h
index aeeb770ba7..a03447f977 100644
--- a/memory/mozalloc/throw_gcc.h
+++ b/memory/mozalloc/throw_gcc.h
@@ -74,50 +74,66 @@ __throw_bad_function_call(void) {
   mozalloc_abort("fatal: STL threw bad_function_call");
 }
 
+#if !defined(_LIBCPP_VERSION)
 MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void __throw_logic_error(
     const char* msg) {
   mozalloc_abort(msg);
 }
+#endif  // _LIBCPP_VERSION
 
+#if !defined(_LIBCPP_VERSION)
 MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void __throw_domain_error(
     const char* msg) {
   mozalloc_abort(msg);
 }
+#endif  // _LIBCPP_VERSION
 
+#if !defined(_LIBCPP_VERSION)
 MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void
 __throw_invalid_argument(const char* msg) {
   mozalloc_abort(msg);
 }
+#endif  // _LIBCPP_VERSION
 
+#if !defined(_LIBCPP_VERSION)
 MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void __throw_length_error(
     const char* msg) {
   mozalloc_abort(msg);
 }
+#endif  // _LIBCPP_VERSION
 
+#if !defined(_LIBCPP_VERSION)
 MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void __throw_out_of_range(
     const char* msg) {
   mozalloc_abort(msg);
 }
+#endif  // _LIBCPP_VERSION
 
 MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void __throw_runtime_error(
     const char* msg) {
   mozalloc_abort(msg);
 }
 
+#if !defined(_LIBCPP_VERSION)
 MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void __throw_range_error(
     const char* msg) {
   mozalloc_abort(msg);
 }
+#endif  // _LIBCPP_VERSION
 
+#if !defined(_LIBCPP_VERSION)
 MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void
 __throw_overflow_error(const char* msg) {
   mozalloc_abort(msg);
 }
+#endif  // _LIBCPP_VERSION
 
+#if !defined(_LIBCPP_VERSION)
 MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void
 __throw_underflow_error(const char* msg) {
   mozalloc_abort(msg);
 }
+#endif  // _LIBCPP_VERSION
 
 MOZ_THROW_NORETURN MOZ_THROW_EXPORT MOZ_THROW_INLINE void __throw_ios_failure(
     const char* msg) {
-- 
2.40.0

