From 1e36ef52696fa282e4a3183344edb9754a51245d Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Tue, 4 Apr 2023 19:07:27 +0300
Subject: [PATCH 04/15] use thinlto for rust.

---
 config/makefiles/rust.mk | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/config/makefiles/rust.mk b/config/makefiles/rust.mk
index 602ba45388..73b7dc2d47 100644
--- a/config/makefiles/rust.mk
+++ b/config/makefiles/rust.mk
@@ -92,7 +92,7 @@ ifndef rustflags_sancov
 # Never enable when coverage is enabled to work around https://github.com/rust-lang/rust/issues/90045.
 ifndef MOZ_CODE_COVERAGE
 ifeq (,$(findstring gkrust_gtest,$(RUST_LIBRARY_FILE)))
-cargo_rustc_flags += -Clto
+cargo_rustc_flags += -Clto=thin
 endif
 # We need -Cembed-bitcode=yes for all crates when using -Clto.
 RUSTFLAGS += -Cembed-bitcode=yes
-- 
2.40.0

