From e752c53a346d81569ec37e7ef2ace462dcbdda4e Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Thu, 7 Jul 2022 23:20:05 +0300
Subject: [PATCH 08/16] cbindgen >=0.24 fixes.

---
 gfx/webrender_bindings/webrender_ffi.h | 2 --
 1 file changed, 2 deletions(-)

diff --git a/gfx/webrender_bindings/webrender_ffi.h b/gfx/webrender_bindings/webrender_ffi.h
index b1d67b17a4..eb79974bdf 100644
--- a/gfx/webrender_bindings/webrender_ffi.h
+++ b/gfx/webrender_bindings/webrender_ffi.h
@@ -73,8 +73,6 @@ struct WrPipelineInfo;
 struct WrPipelineIdAndEpoch;
 using WrPipelineIdEpochs = nsTArray<WrPipelineIdAndEpoch>;
 
-const uint64_t ROOT_CLIP_CHAIN = ~0;
-
 }  // namespace wr
 }  // namespace mozilla
 
-- 
2.37.0

