From c81eb951c92bc7d71de1fdecbe6fd43da379725d Mon Sep 17 00:00:00 2001
From: Johannes <johannes.brechtmann@gmail.com>
Date: Fri, 1 May 2020 17:20:29 +0200
Subject: [PATCH 18/44] musl: make SYS_fork non-fatal, musl uses it for fork(2)

Signed-off-by: Thomas Deutschmann <whissi@gentoo.org>
---
 security/sandbox/linux/SandboxFilter.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/security/sandbox/linux/SandboxFilter.cpp b/security/sandbox/linux/SandboxFilter.cpp
index b60902e841..98a17f40f8 100644
--- a/security/sandbox/linux/SandboxFilter.cpp
+++ b/security/sandbox/linux/SandboxFilter.cpp
@@ -1310,6 +1310,10 @@ class ContentSandboxPolicy : public SandboxPolicyCommon {
         // usually do something reasonable on error.
       case __NR_clone:
         return ClonePolicy(Error(EPERM));
+#  ifdef __NR_fork
+      case __NR_fork:
+        return Error(ENOSYS);
+#  endif
 
 #  ifdef __NR_fadvise64
       case __NR_fadvise64:
-- 
2.31.1

