From 287b2bbe7cb901669e6cb24ba6787a4b49a190d9 Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Fri, 8 Jul 2022 01:52:43 +0300
Subject: [PATCH 15/16] nuke ct data reporting.

---
 browser/moz.configure | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/browser/moz.configure b/browser/moz.configure
index 8653bcbb16..935af4065a 100644
--- a/browser/moz.configure
+++ b/browser/moz.configure
@@ -5,11 +5,11 @@
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
 imply_option("MOZ_PLACES", True)
-imply_option("MOZ_SERVICES_HEALTHREPORT", True)
+imply_option("MOZ_SERVICES_HEALTHREPORT", False)
 imply_option("MOZ_SERVICES_SYNC", True)
 imply_option("MOZ_DEDICATED_PROFILES", True)
 imply_option("MOZ_BLOCK_PROFILE_DOWNGRADE", True)
-imply_option("MOZ_NORMANDY", True)
+imply_option("MOZ_NORMANDY", False)
 
 with only_when(target_is_linux & compile_environment):
     option(env="MOZ_NO_PIE_COMPAT", help="Enable non-PIE wrapper")
-- 
2.37.0

