From 44db4b702cfbede9c55d1d53cbe691770a9e71fa Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Mon, 26 Sep 2022 10:25:50 +0300
Subject: [PATCH 15/15] fix dbus.

---
 toolkit/moz.configure | 32 ++++++++++++++++++++------------
 1 file changed, 20 insertions(+), 12 deletions(-)

diff --git a/toolkit/moz.configure b/toolkit/moz.configure
index c6b0858324..c8ae802c8c 100644
--- a/toolkit/moz.configure
+++ b/toolkit/moz.configure
@@ -2362,18 +2362,6 @@ with only_when(compile_environment & target_is_windows):
     )
     set_config("MOZ_D3DCOMPILER_VISTA_DLL_PATH", d3d_compiler_dll.path)
 
-# Remoting protocol support
-# ==============================================================
-
-
-@depends(toolkit)
-def has_remote(toolkit):
-    if toolkit in ("gtk", "windows", "cocoa"):
-        return True
-
-
-set_config("MOZ_HAS_REMOTE", has_remote)
-set_define("MOZ_HAS_REMOTE", has_remote)
 
 # RLBox Library Sandboxing wasm support
 # ==============================================================
@@ -2909,6 +2897,26 @@ with only_when(toolkit_gtk):
         set_config("MOZ_ENABLE_DBUS", True)
         set_define("MOZ_ENABLE_DBUS", True)
 
+# Remoting protocol support
+# ==============================================================
+
+@depends(
+    toolkit,
+    depends("--enable-dbus", when=toolkit_gtk)(lambda x: x),
+    depends(wayland_headers, when=toolkit_gtk)(lambda x: x),
+    depends(x11_headers, when=toolkit_gtk)(lambda x: x),
+)
+def has_remote(toolkit, dbus, wayland, x11):
+    if toolkit in ("windows", "cocoa"):
+        return True
+    if toolkit == "gtk":
+        if not dbus and wayland and not x11:
+            return
+        return True
+
+
+set_config("MOZ_HAS_REMOTE", has_remote)
+set_define("MOZ_HAS_REMOTE", has_remote)
 
 # Necko's wifi scanner
 # ==============================================================
-- 
2.37.1

