diff -uraN a/config/system-headers.mozbuild b/config/system-headers.mozbuild
--- a/config/system-headers.mozbuild	2022-03-13 17:31:36.000000000 +0300
+++ b/config/system-headers.mozbuild	2022-03-26 16:05:20.279997983 +0300
@@ -269,8 +269,6 @@
     'freetype/tttables.h',
     'fribidi/fribidi.h',
     'FSp_fopen.h',
-    'fstream',
-    'fstream.h',
     'ft2build.h',
     'ftadvanc.h',
     'ftbitmap.h',
