From 9fa200270558fa30fdd65c8395fcf380d0d7b7d0 Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Fri, 1 Apr 2022 18:09:30 +0300
Subject: [PATCH 2/8] libcxx: force link against compiler-rt.

---
 libcxx/src/CMakeLists.txt | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/libcxx/src/CMakeLists.txt b/libcxx/src/CMakeLists.txt
index 12dcdf954..be6652f08 100644
--- a/libcxx/src/CMakeLists.txt
+++ b/libcxx/src/CMakeLists.txt
@@ -213,6 +213,11 @@ if (LIBCXX_ENABLE_SHARED)
   target_include_directories(cxx_shared PRIVATE ${CMAKE_CURRENT_SOURCE_DIR})
   target_link_libraries(cxx_shared PUBLIC cxx-headers
                                    PRIVATE ${LIBCXX_LIBRARIES})
+
+  if (LIBCXX_USE_COMPILER_RT)
+    target_link_libraries(cxx_shared PRIVATE ${LIBCXX_BUILTINS_LIBRARY})
+  endif()
+
   set_target_properties(cxx_shared
     PROPERTIES
       COMPILE_FLAGS "${LIBCXX_COMPILE_FLAGS}"
@@ -291,6 +296,11 @@ if (LIBCXX_ENABLE_STATIC)
   target_include_directories(cxx_static PRIVATE ${CMAKE_CURRENT_SOURCE_DIR})
   target_link_libraries(cxx_static PUBLIC cxx-headers
                                    PRIVATE ${LIBCXX_LIBRARIES})
+
+  if (LIBCXX_USE_COMPILER_RT)
+    target_link_libraries(cxx_static PRIVATE ${LIBCXX_BUILTINS_LIBRARY})
+  endif()
+
   set_target_properties(cxx_static
     PROPERTIES
       COMPILE_FLAGS "${LIBCXX_COMPILE_FLAGS}"
-- 
2.35.1

