diff --git a/cmake/config-ix.cmake b/cmake/config-ix.cmake
index 15b5208..2bce4f1 100644
--- a/libcxxabi/cmake/config-ix.cmake
+++ b/libcxxabi/cmake/config-ix.cmake
@@ -84,7 +84,5 @@ if(FUCHSIA)
 else()
   check_library_exists(dl dladdr "" LIBCXXABI_HAS_DL_LIB)
   check_library_exists(pthread pthread_once "" LIBCXXABI_HAS_PTHREAD_LIB)
-  check_library_exists(c __cxa_thread_atexit_impl ""
-    LIBCXXABI_HAS_CXA_THREAD_ATEXIT_IMPL)
   check_library_exists(System write "" LIBCXXABI_HAS_SYSTEM_LIB)
 endif()
-- 
2.29.2

