From ea514b599856026a8ec5015c5fa03bd3f4f7579e Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Sun, 29 Aug 2021 07:36:02 +0300
Subject: [PATCH 3/7] libcxxabi: don't link against __cxa_thread_atexit_im.

---
 libcxxabi/cmake/config-ix.cmake | 2 --
 1 file changed, 2 deletions(-)

diff --git a/libcxxabi/cmake/config-ix.cmake b/libcxxabi/cmake/config-ix.cmake
index 7f1cecbcd..f75ecfc62 100644
--- a/libcxxabi/cmake/config-ix.cmake
+++ b/libcxxabi/cmake/config-ix.cmake
@@ -94,7 +94,5 @@ if(FUCHSIA)
 else()
   check_library_exists(dl dladdr "" LIBCXXABI_HAS_DL_LIB)
   check_library_exists(pthread pthread_once "" LIBCXXABI_HAS_PTHREAD_LIB)
-  check_library_exists(c __cxa_thread_atexit_impl ""
-    LIBCXXABI_HAS_CXA_THREAD_ATEXIT_IMPL)
   check_library_exists(System write "" LIBCXXABI_HAS_SYSTEM_LIB)
 endif()
-- 
2.33.0

