From 01af1555aeef13825f4c6dd24b21dd6aecc6b922 Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Thu, 28 Apr 2022 13:15:04 +0300
Subject: [PATCH 3/8] libcxxabi: don't link against __cxa_thread_atexit_im.

---
 libcxxabi/cmake/config-ix.cmake | 2 --
 1 file changed, 2 deletions(-)

diff --git a/libcxxabi/cmake/config-ix.cmake b/libcxxabi/cmake/config-ix.cmake
index efc572ea7..e09d32734 100644
--- a/libcxxabi/cmake/config-ix.cmake
+++ b/libcxxabi/cmake/config-ix.cmake
@@ -95,7 +95,5 @@ if(FUCHSIA)
 else()
   check_library_exists(dl dladdr "" LIBCXXABI_HAS_DL_LIB)
   check_library_exists(pthread pthread_once "" LIBCXXABI_HAS_PTHREAD_LIB)
-  check_library_exists(c __cxa_thread_atexit_impl ""
-    LIBCXXABI_HAS_CXA_THREAD_ATEXIT_IMPL)
   check_library_exists(System write "" LIBCXXABI_HAS_SYSTEM_LIB)
 endif()
-- 
2.36.0

