From 3c57961279a0703dfa97e4ae076ab543052f6bf0 Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Sun, 19 Mar 2023 05:23:31 +0300
Subject: [PATCH 05/10] llvm: do not enforce CMAKE_BUILD_TYPE.

---
 llvm/CMakeLists.txt | 12 ------------
 1 file changed, 12 deletions(-)

diff --git a/llvm/CMakeLists.txt b/llvm/CMakeLists.txt
index b52243977..f2b3d7311 100644
--- a/llvm/CMakeLists.txt
+++ b/llvm/CMakeLists.txt
@@ -93,18 +93,6 @@ else()
   set(CMAKE_CXX_EXTENSIONS NO)
 endif()
 
-if (NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
-  message(FATAL_ERROR "
-No build type selected. You need to pass -DCMAKE_BUILD_TYPE=<type> in order to configure LLVM.
-Available options are:
-  * -DCMAKE_BUILD_TYPE=Release - For an optimized build with no assertions or debug info.
-  * -DCMAKE_BUILD_TYPE=Debug - For an unoptimized build with assertions and debug info.
-  * -DCMAKE_BUILD_TYPE=RelWithDebInfo - For an optimized build with no assertions but with debug info.
-  * -DCMAKE_BUILD_TYPE=MinSizeRel - For a build optimized for size instead of speed.
-Learn more about these options in our documentation at https://llvm.org/docs/CMake.html#cmake-build-type
-")
-endif()
-
 # Set default build type for cmake's try_compile module.
 # CMake 3.17 or newer sets CMAKE_DEFAULT_BUILD_TYPE to one of the
 # items from CMAKE_CONFIGURATION_TYPES. Logic below can be further
-- 
2.39.2

