From 950e44fd68ab48c8aa94dc23c0bded9452aa392f Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Sun, 19 Mar 2023 16:03:54 +0300
Subject: [PATCH 06/11] clang: disable multiarch layout.

---
 clang/lib/Driver/ToolChains/Linux.cpp | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/clang/lib/Driver/ToolChains/Linux.cpp b/clang/lib/Driver/ToolChains/Linux.cpp
index c6fb290ff..eeb656790 100644
--- a/clang/lib/Driver/ToolChains/Linux.cpp
+++ b/clang/lib/Driver/ToolChains/Linux.cpp
@@ -47,6 +47,11 @@ std::string Linux::getMultiarchTriple(const Driver &D,
   bool IsMipsR6 = TargetTriple.getSubArch() == llvm::Triple::MipsSubArch_r6;
   bool IsMipsN32Abi = TargetTriple.getEnvironment() == llvm::Triple::GNUABIN32;
 
+  /* we are not multiarch */
+  if (TargetTriple.isMusl()) {
+    return "";
+  }
+
   // For most architectures, just use whatever we have rather than trying to be
   // clever.
   switch (TargetTriple.getArch()) {
-- 
2.39.2

