From 5c1e4b9330d8126bb7ede992567a1de19e64c864 Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Thu, 24 Mar 2022 12:44:31 +0300
Subject: [PATCH 6/7] llvm: don't build llvm-locstats.

---
 llvm/CMakeLists.txt | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/llvm/CMakeLists.txt b/llvm/CMakeLists.txt
index 00d01f4dd..785306fc8 100644
--- a/llvm/CMakeLists.txt
+++ b/llvm/CMakeLists.txt
@@ -1257,7 +1257,3 @@ if (LLVM_INCLUDE_BENCHMARKS)
     ${CMAKE_CURRENT_BINARY_DIR}/third-party/benchmark)
   add_subdirectory(benchmarks)
 endif()
-
-if (LLVM_INCLUDE_UTILS AND LLVM_INCLUDE_TOOLS)
-  add_subdirectory(utils/llvm-locstats)
-endif()
-- 
2.35.1

