From d38043720e87b66723492be42d8eaa83311aacb1 Mon Sep 17 00:00:00 2001
From: Thomas Deutschmann <whissi@gentoo.org>
Date: Sat, 10 Oct 2020 17:13:46 +0200
Subject: [PATCH 39/40] bmo#1669442: Fix dependency between libxul and mozgtk
 after bug 1573566

Signed-off-by: Thomas Deutschmann <whissi@gentoo.org>
---
 config/recurse.mk | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/config/recurse.mk b/config/recurse.mk
index a1567ae3cd..7396c72716 100644
--- a/config/recurse.mk
+++ b/config/recurse.mk
@@ -206,7 +206,7 @@ endif
 # Interdependencies that moz.build world don't know about yet for compilation.
 # Note some others are hardcoded or "guessed" in recursivemake.py and emitter.py
 ifeq ($(MOZ_WIDGET_TOOLKIT),gtk)
-toolkit/library/target: widget/gtk/mozgtk/gtk3/target
+toolkit/library/build/target: widget/gtk/mozgtk/gtk3/target
 endif
 
 ifndef MOZ_FOLD_LIBS
-- 
2.28.0

