diff -uraN a/kernel/panic.c b/kernel/panic.c
--- a/kernel/panic.c	2019-09-06 13:42:01.000000000 +0300
+++ b/kernel/panic.c	2019-09-08 16:32:30.491660999 +0300
@@ -210,7 +210,7 @@
 	if (len && buf[len - 1] == '\n')
 		buf[len - 1] = '\0';
 
-	pr_emerg("Kernel panic - not syncing: %s\n", buf);
+	pr_emerg("oh god oh fuck - not syncing: %s\n", buf);
 #ifdef CONFIG_DEBUG_BUGVERBOSE
 	/*
 	 * Avoid nested stack-dumping if a panic occurs during oops processing
@@ -327,7 +327,7 @@
 #if defined(CONFIG_S390)
 	disabled_wait();
 #endif
-	pr_emerg("---[ end Kernel panic - not syncing: %s ]---\n", buf);
+	pr_emerg("---[ end the panic thingy - not syncing: %s ]---\n", buf);
 
 	/* Do not scroll important messages printed above */
 	suppress_printk = 1;
