diff -uraN a/neo/CMakeLists.txt b/neo/CMakeLists.txt
--- a/neo/CMakeLists.txt	2021-03-14 05:51:07.000000000 +0300
+++ b/neo/CMakeLists.txt	2021-05-31 00:47:38.028197416 +0300
@@ -347,30 +347,9 @@
 	)
 endif()
 
-# fallback for cmake versions without GNUInstallDirs
-if(GNUINSTALLDIRS MATCHES "NOTFOUND")
-	set(CMAKE_INSTALL_BINDIR		"bin"
-		CACHE PATH "user executables (bin)")
-	set(CMAKE_INSTALL_LIBDIR		"lib${LIB_SUFFIX}"
-		CACHE PATH "object code libraries (lib${LIB_SUFFIX})")
-	set(CMAKE_INSTALL_DATAROOTDIR	"share"
-		CACHE PATH "read-only architecture-independent data root (share)")
-	set(CMAKE_INSTALL_DATADIR		"${CMAKE_INSTALL_DATAROOTDIR}"
-		CACHE PATH "read-only architecture-independent data (DATAROOTDIR)")
-
-	mark_as_advanced(CMAKE_INSTALL_BINDIR CMAKE_INSTALL_LIBDIR CMAKE_INSTALL_DATAROOTDIR CMAKE_INSTALL_DATADIR)
-	foreach(dir BINDIR LIBDIR DATAROOTDIR DATADIR)
-		if(NOT IS_ABSOLUTE ${CMAKE_INSTALL_${dir}})
-			set(CMAKE_INSTALL_FULL_${dir} "${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_${dir}}")
-		else()
-			set(CMAKE_INSTALL_FULL_${dir} "${CMAKE_INSTALL_${dir}}")
-		endif()
-	endforeach()
-endif()
-
-set(bindir		"${CMAKE_INSTALL_FULL_BINDIR}")
-set(libdir		"${CMAKE_INSTALL_FULL_LIBDIR}/dhewm3")
-set(datadir		"${CMAKE_INSTALL_FULL_DATADIR}/dhewm3")
+set(bindir		"${CMAKE_INSTALL_PREFIX}")
+set(libdir		"${CMAKE_INSTALL_PREFIX}/lib")
+set(datadir		"${CMAKE_INSTALL_PREFIX}/data")
 
 configure_file(
 	"${CMAKE_SOURCE_DIR}/config.h.in"
