From 352ffda37998506617f169e6e5af76db5debd732 Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Wed, 12 Apr 2023 01:45:15 +0300
Subject: [PATCH 3/3] do not build native.

---
 src/include/defaults.mk | 8 +-------
 1 file changed, 1 insertion(+), 7 deletions(-)

diff --git a/src/include/defaults.mk b/src/include/defaults.mk
index 6874c9f..2b3f960 100644
--- a/src/include/defaults.mk
+++ b/src/include/defaults.mk
@@ -71,16 +71,10 @@ override SOFLAGS = $(_SOFLAGS) \
 		   -Wl,--version-script=$(MAP) \
 		   $(call family,SOFLAGS)
 
-HOST_ARCH=$(shell uname -m)
-ifneq ($(HOST_ARCH),ia64)
-	HOST_MARCH=-march=native
-else
-	HOST_MARCH=
-endif
 HOST_CPPFLAGS ?= $(CPPFLAGS)
 override _HOST_CPPFLAGS := $(HOST_CPPFLAGS)
 override HOST_CPPFLAGS = $(_HOST_CPPFLAGS) \
-			 -DEFIVAR_BUILD_ENVIRONMENT $(HOST_MARCH)
+			 -DEFIVAR_BUILD_ENVIRONMENT
 HOST_CFLAGS_GCC ?=
 HOST_CFLAGS_CLANG ?=
 HOST_CFLAGS ?= $(CFLAGS) $(call family,HOST_CFLAGS)
-- 
2.40.0

