diff -uraN a/Makefile b/Makefile
--- a/Makefile	2019-11-18 01:47:30.000000000 +0300
+++ b/Makefile	2019-11-18 04:38:35.257863228 +0300
@@ -702,6 +702,8 @@
 
 ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE
 KBUILD_CFLAGS += -O2
+else ifdef CONFIG_CC_OPTIMIZE_HARDER
+KBUILD_CFLAGS  += -O3
 else ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE_O3
 KBUILD_CFLAGS += -O3
 else ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
diff -uraN a/init/Kconfig b/init/Kconfig
--- a/init/Kconfig	2019-11-18 01:47:30.000000000 +0300
+++ b/init/Kconfig	2019-11-18 04:39:31.562864138 +0300
@@ -1242,6 +1242,13 @@
 	  with the "-O2" compiler flag for best performance and most
 	  helpful compile-time warnings.
 
+config CC_OPTIMIZE_HARDER
+       bool "Optimize harder"
+       help
+         This option will pass "-O3" to your compiler resulting in a
+         larger and faster kernel. The more complex optimizations also
+         increase compilation time and may affect stability.
+
 config CC_OPTIMIZE_FOR_PERFORMANCE_O3
 	bool "Optimize more for performance (-O3)"
 	depends on ARC
