From 7ea90a6c3fd1dd36a349df294fc78f716e1976cb Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Mon, 11 Oct 2021 15:23:35 +0300
Subject: [PATCH 2/4] install: fix chown.

---
 coreutils/install.c | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/coreutils/install.c b/coreutils/install.c
index c0f1c53..a81a5a1 100644
--- a/coreutils/install.c
+++ b/coreutils/install.c
@@ -244,6 +244,14 @@ int install_main(int argc, char **argv)
 			}
 		}
 
+		/* Set the user and group id */
+		if ((opts & (OPT_OWNER|OPT_GROUP))
+		 && lchown(dest, uid, gid) == -1
+		) {
+			bb_perror_msg("can't change %s of %s", "ownership", dest);
+			ret = EXIT_FAILURE;
+		}
+
 		/* Set the file mode (always, not only with -m).
 		 * GNU coreutils 6.10 is not affected by umask. */
 		if (chmod(dest, mode) == -1) {
@@ -254,13 +262,6 @@ int install_main(int argc, char **argv)
 		if (use_default_selinux_context)
 			setdefaultfilecon(dest);
 #endif
-		/* Set the user and group id */
-		if ((opts & (OPT_OWNER|OPT_GROUP))
-		 && lchown(dest, uid, gid) == -1
-		) {
-			bb_perror_msg("can't change %s of %s", "ownership", dest);
-			ret = EXIT_FAILURE;
-		}
  next:
 		if (ENABLE_FEATURE_CLEAN_UP && isdir)
 			free(dest);
-- 
2.33.0

