From 5595b8935719b6ad1e117d4e36524a97f572cfd5 Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Tue, 3 Jan 2023 20:47:27 +0300
Subject: [PATCH 4/5] unzip: exit with 0 after printing usage.

---
 archival/unzip.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/archival/unzip.c b/archival/unzip.c
index b27dd21..31a16c1 100644
--- a/archival/unzip.c
+++ b/archival/unzip.c
@@ -618,6 +618,7 @@ int unzip_main(int argc, char **argv)
 			break;
 
 		default:
+			xfunc_error_retval = 0;
 			bb_show_usage();
 		}
 	}
@@ -641,6 +642,7 @@ int unzip_main(int argc, char **argv)
 #endif
 
 	if (!src_fn) {
+		xfunc_error_retval = 0;
 		bb_show_usage();
 	}
 
-- 
2.39.0

