From 6ee55bf3860e6e128e2836624e5285cb53f66567 Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Mon, 30 Jan 2023 21:31:34 +0300
Subject: [PATCH] socket.c: fix musl BUFSIZ.

---
 socket.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/socket.c b/socket.c
index 352c8f3..ddc7461 100644
--- a/socket.c
+++ b/socket.c
@@ -35,6 +35,8 @@
 # include <err.h>
 #endif
 
+#include <stdio.h>
+
 #include "extern.h"
 
 /*
-- 
2.39.1

