diff -uraN a/libtransmission/session.h b/libtransmission/session.h
--- a/libtransmission/session.h	2018-05-01 20:21:08.228549615 +0300
+++ b/libtransmission/session.h	2019-12-20 07:51:59.014307327 +0300
@@ -14,7 +14,7 @@
 #ifndef TR_INTERNAL_H
 #define TR_INTERNAL_H 1
 
-#define TR_NAME "Transmission"
+#define TR_NAME "Deluge"
 
 #ifndef UNUSED
  #ifdef __GNUC__
diff -uraN a/libtransmission/version.h b/libtransmission/version.h
--- a/libtransmission/version.h	2018-05-01 20:21:24.159620418 +0300
+++ b/libtransmission/version.h	2019-12-20 07:51:04.103306439 +0300
@@ -1,12 +1,12 @@
 #pragma once
 
-#define PEERID_PREFIX             "-TR2940-"
-#define USERAGENT_PREFIX          "2.94"
+#define PEERID_PREFIX             "-DE24CD-"
+#define USERAGENT_PREFIX          "2.4.12b2.dev3"
 #define VCS_REVISION              "d8e60ee44f"
 #define VCS_REVISION_NUM          d8e60ee44f
-#define SHORT_VERSION_STRING      "2.94"
-#define LONG_VERSION_STRING       "2.94 (d8e60ee44f)"
-#define VERSION_STRING_INFOPLIST  2.94
+#define SHORT_VERSION_STRING      "2.4.12b2.dev3"
+#define LONG_VERSION_STRING       "2.4.12b2.dev3"
+#define VERSION_STRING_INFOPLIST  "2.4.12b2.dev3"
 #define BUILD_STRING_INFOPLIST    14714.2.94
 #define MAJOR_VERSION             2
 #define MINOR_VERSION             94
