From f2259be4f0c67e1f36a3ed860e0c655e81df8fdf Mon Sep 17 00:00:00 2001
From: mssx86 <mssftw@protonmail.com>
Date: Sun, 7 Jul 2019 15:44:31 +0300
Subject: added headers required for building with glibc.

---
 libutil/tty.c | 1 +
 mknod.c       | 1 +
 mountpoint.c  | 1 +
 stat.c        | 3 ++-
 4 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/libutil/tty.c b/libutil/tty.c
index bceb01e..73e466b 100644
--- a/libutil/tty.c
+++ b/libutil/tty.c
@@ -11,6 +11,7 @@
 #include <unistd.h>
 
 #include "../util.h"
+#include <sys/sysmacros.h>
 
 void
 devtotty(int dev, int *tty_maj, int *tty_min)
diff --git a/mknod.c b/mknod.c
index 8de35c7..702d6c1 100644
--- a/mknod.c
+++ b/mknod.c
@@ -9,6 +9,7 @@
 #include <unistd.h>
 
 #include "util.h"
+#include <sys/sysmacros.h>
 
 static void
 usage(void)
diff --git a/mountpoint.c b/mountpoint.c
index 8f205a2..44ffbac 100644
--- a/mountpoint.c
+++ b/mountpoint.c
@@ -9,6 +9,7 @@
 #include <unistd.h>
 
 #include "util.h"
+#include <sys/sysmacros.h>
 
 static void
 usage(void)
diff --git a/stat.c b/stat.c
index 220a659..74d9c05 100644
--- a/stat.c
+++ b/stat.c
@@ -9,6 +9,7 @@
 #include <unistd.h>
 
 #include "util.h"
+#include <sys/sysmacros.h>
 
 static void
 show_stat_terse(const char *file, struct stat *st)
@@ -86,4 +87,4 @@ main(int argc, char *argv[])
 	}
 
 	return ret;
-}
\ No newline at end of file
+}
-- 
2.20.1

