From 9f6a9a6a94ceb956923b02d4ccb16187810749a3 Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Wed, 1 Mar 2023 03:13:57 +0300
Subject: [PATCH] link against lltdl.

---
 src/meson.build | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/meson.build b/src/meson.build
index d480475..c09c94e 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -26,6 +26,7 @@ executable('pavucontrol',
   pavucontrol_sources,
   install: true,
   cpp_args: pavucontrol_defines,
+  link_args: '-lltdl',
   include_directories : configinc,
   dependencies : pavucontrol_deps,
 )
-- 
2.39.2

