From 17f05c699cf4949f876b110830b250af009b2938 Mon Sep 17 00:00:00 2001
From: mssx86 <mssftw@protonmail.com>
Date: Sun, 7 Jul 2019 15:45:51 +0300
Subject: config.mk customization.

---
 config.mk | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/config.mk b/config.mk
index 245a590..0c998d2 100644
--- a/config.mk
+++ b/config.mk
@@ -2,10 +2,10 @@
 VERSION = 0.1
 
 # paths
-PREFIX = /usr/local
+PREFIX = /opt/suckless
 MANPREFIX = $(PREFIX)/share/man
 
-CC = cc
+CC = clang
 AR = ar
 RANLIB = ranlib
 
-- 
2.20.1

