From 704e00270fa8ce5a541cd83401681bd7ea5725a9 Mon Sep 17 00:00:00 2001
From: mssx86 <mssftw@protonmail.com>
Date: Sun, 7 Jul 2019 15:49:36 +0300
Subject: config.mk customization.

---
 config.mk | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/config.mk b/config.mk
index 9fb18da..ebfa6fb 100644
--- a/config.mk
+++ b/config.mk
@@ -2,10 +2,10 @@
 VERSION = 0.0
 
 # paths
-PREFIX = /usr/local
+PREFIX = /opt/suckless
 MANPREFIX = $(PREFIX)/share/man
 
-CC = cc
+CC = clang
 AR = ar
 RANLIB = ranlib
 
-- 
2.20.1

