diff -uraN a/kernel/panic.c b/kernel/panic.c
--- a/kernel/panic.c	2019-12-21 12:35:48.000000000 +0300
+++ b/kernel/panic.c	2019-12-29 21:00:41.834934885 +0300
@@ -104,7 +104,7 @@
 	va_start(args, fmt);
 	vsnprintf(buf, sizeof(buf), fmt, args);
 	va_end(args);
-	pr_emerg("Kernel panic - not syncing: %s\n", buf);
+	pr_emerg("oh god oh fuck - not syncing: %s\n", buf);
 #ifdef CONFIG_DEBUG_BUGVERBOSE
 	/*
 	 * Avoid nested stack-dumping if a panic occurs during oops processing
