From cf16c6c2fd16ed7712c3de0cd13a30e798d1e8b2 Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Sun, 2 Apr 2023 07:45:57 +0300
Subject: [PATCH 2/6] libunwind adjustments.

---
 library/unwind/src/lib.rs | 18 ++----------------
 1 file changed, 2 insertions(+), 16 deletions(-)

diff --git a/library/unwind/src/lib.rs b/library/unwind/src/lib.rs
index b655bae96..1e82df164 100644
--- a/library/unwind/src/lib.rs
+++ b/library/unwind/src/lib.rs
@@ -37,22 +37,8 @@ cfg_if::cfg_if! {
 }
 
 #[cfg(target_env = "musl")]
-cfg_if::cfg_if! {
-    if #[cfg(all(feature = "llvm-libunwind", feature = "system-llvm-libunwind"))] {
-        compile_error!("`llvm-libunwind` and `system-llvm-libunwind` cannot be enabled at the same time");
-    } else if #[cfg(feature = "llvm-libunwind")] {
-        #[link(name = "unwind", kind = "static", modifiers = "-bundle")]
-        extern "C" {}
-    } else if #[cfg(feature = "system-llvm-libunwind")] {
-        #[link(name = "unwind", kind = "static", modifiers = "-bundle", cfg(target_feature = "crt-static"))]
-        #[link(name = "unwind", cfg(not(target_feature = "crt-static")))]
-        extern "C" {}
-    } else {
-        #[link(name = "unwind", kind = "static", modifiers = "-bundle", cfg(target_feature = "crt-static"))]
-        #[link(name = "gcc_s", cfg(not(target_feature = "crt-static")))]
-        extern "C" {}
-    }
-}
+#[link(name = "unwind")]
+extern "C" {}
 
 // This is the same as musl except that we default to using the system libunwind
 // instead of libgcc.
-- 
2.40.0

