From c0d12e1778d4a0d039fcd1673957c792d4779d4f Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Sat, 4 Dec 2021 15:56:41 +0300
Subject: [PATCH 4/6] use cpp to compile cpp code on musl.

---
 src/bootstrap/native.rs | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/bootstrap/native.rs b/src/bootstrap/native.rs
index 27c9bb250..a67d1bf21 100644
--- a/src/bootstrap/native.rs
+++ b/src/bootstrap/native.rs
@@ -1056,7 +1056,7 @@ impl Step for Libunwind {
             cpp_cfg.flag("-std=c++11");
         }
 
-        if self.target.contains("x86_64-fortanix-unknown-sgx") || self.target.contains("musl") {
+        if self.target.contains("x86_64-fortanix-unknown-sgx") {
             // use the same GCC C compiler command to compile C++ code so we do not need to setup the
             // C++ compiler env variables on the builders.
             // Don't set this for clang++, as clang++ is able to compile this without libc++.
-- 
2.31.1

