From abcd356b5223061c8d4e1003cd0eaa0e31eadd6a Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Fri, 27 Jan 2023 10:52:47 +0300
Subject: [PATCH 4/6] use c++ to compile c++ code on musl.

---
 src/bootstrap/native.rs | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/bootstrap/native.rs b/src/bootstrap/native.rs
index f6c453ebe..d34d4d06c 100644
--- a/src/bootstrap/native.rs
+++ b/src/bootstrap/native.rs
@@ -1312,7 +1312,7 @@ impl Step for Libunwind {
             cpp_cfg.flag("-std=c++11");
         }
 
-        if self.target.contains("x86_64-fortanix-unknown-sgx") || self.target.contains("musl") {
+        if self.target.contains("x86_64-fortanix-unknown-sgx") {
             // use the same GCC C compiler command to compile C++ code so we do not need to setup the
             // C++ compiler env variables on the builders.
             // Don't set this for clang++, as clang++ is able to compile this without libc++.
-- 
2.39.0

