From c4917416fb86acb39344c9ecf3e7da64aa4a9625 Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Sat, 23 Oct 2021 00:31:05 +0300
Subject: [PATCH 4/6] use cpp to compile cpp code on musl.

---
 src/bootstrap/native.rs | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/bootstrap/native.rs b/src/bootstrap/native.rs
index 8cf43eac4..f416cb211 100644
--- a/src/bootstrap/native.rs
+++ b/src/bootstrap/native.rs
@@ -1051,7 +1051,7 @@ impl Step for Libunwind {
             cpp_cfg.flag("-std=c++11");
         }
 
-        if self.target.contains("x86_64-fortanix-unknown-sgx") || self.target.contains("musl") {
+        if self.target.contains("x86_64-fortanix-unknown-sgx") {
             // use the same GCC C compiler command to compile C++ code so we do not need to setup the
             // C++ compiler env variables on the builders.
             // Don't set this for clang++, as clang++ is able to compile this without libc++.
-- 
2.31.1

