From ec203dfd034e2b984a5f60bf0fad24c36b66b3a1 Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Sun, 2 Apr 2023 07:47:12 +0300
Subject: [PATCH 4/6] use c++ to compile c++ code on musl.

---
 src/bootstrap/llvm.rs | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/bootstrap/llvm.rs b/src/bootstrap/llvm.rs
index cc2b45a9b..821afba1c 100644
--- a/src/bootstrap/llvm.rs
+++ b/src/bootstrap/llvm.rs
@@ -1235,7 +1235,7 @@ impl Step for Libunwind {
             cpp_cfg.flag("-std=c++11");
         }
 
-        if self.target.contains("x86_64-fortanix-unknown-sgx") || self.target.contains("musl") {
+        if self.target.contains("x86_64-fortanix-unknown-sgx") {
             // use the same GCC C compiler command to compile C++ code so we do not need to setup the
             // C++ compiler env variables on the builders.
             // Don't set this for clang++, as clang++ is able to compile this without libc++.
-- 
2.40.0

