From 96a191d73fc02c8c9842343bdb8d25be7fb41a16 Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Fri, 27 Jan 2023 10:53:19 +0300
Subject: [PATCH 5/6] fix curl cert error with cargo.

---
 vendor/curl-sys/build.rs | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/vendor/curl-sys/build.rs b/vendor/curl-sys/build.rs
index 821959488..44a564a03 100644
--- a/vendor/curl-sys/build.rs
+++ b/vendor/curl-sys/build.rs
@@ -117,6 +117,8 @@ fn main() {
         .include("curl/lib")
         .include("curl/include")
         .define("BUILDING_LIBCURL", None)
+        .define("CURL_CA_BUNDLE", "\"/etc/ssl/cert.pem\"")
+        .define("CURL_CA_PATH", "\"/etc/ssl/certs\"")
         .define("CURL_DISABLE_DICT", None)
         .define("CURL_DISABLE_GOPHER", None)
         .define("CURL_DISABLE_IMAP", None)
-- 
2.39.0

