From 141943bffe5a80831c2281488b2e1d08a54ce978 Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Wed, 25 May 2022 12:08:10 +0300
Subject: [PATCH 5/6] fix curl cert error with cargo.

---
 vendor/curl-sys/build.rs | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/vendor/curl-sys/build.rs b/vendor/curl-sys/build.rs
index 8615c4db3..290897613 100644
--- a/vendor/curl-sys/build.rs
+++ b/vendor/curl-sys/build.rs
@@ -111,6 +111,8 @@ fn main() {
         .include("curl/lib")
         .include("curl/include")
         .define("BUILDING_LIBCURL", None)
+        .define("CURL_CA_BUNDLE", "\"/etc/ssl/cert.pem\"")
+        .define("CURL_CA_PATH", "\"/etc/ssl/certs\"")
         .define("CURL_DISABLE_DICT", None)
         .define("CURL_DISABLE_GOPHER", None)
         .define("CURL_DISABLE_IMAP", None)
-- 
2.36.0

