--- a/src/drmmode_display.h	2023-02-22 20:14:50.000000000 +0300
+++ b/src/drmmode_display.h	2023-04-30 01:01:58.668247272 +0300
@@ -36,6 +36,8 @@
 #include "amdgpu_probe.h"
 #include "amdgpu.h"
 
+#include <X11/extensions/dpmsconst.h>
+
 /*
  * Enum of non-legacy color management properties, according to DRM. Note that
  * the values should be incremental (with the exception of the INVALID member),
