diff -uraN a/build/config/ui.gni b/build/config/ui.gni
--- a/build/config/ui.gni	2023-04-27 00:42:47.838194800 +0300
+++ b/build/config/ui.gni	2023-05-06 18:59:33.922075000 +0300
@@ -41,7 +41,9 @@
 
 assert(!use_glib || (is_linux && !is_castos))
 
-use_atk = use_glib && current_toolchain == default_toolchain
+declare_args() {
+  use_atk = use_glib && current_toolchain == default_toolchain
+}
 
 # Whether using Xvfb to provide a display server for a test might be
 # necessary.
diff -uraN a/chrome/browser/chrome_browser_main_linux.cc b/chrome/browser/chrome_browser_main_linux.cc
--- a/chrome/browser/chrome_browser_main_linux.cc	2023-04-27 00:42:52.806633200 +0300
+++ b/chrome/browser/chrome_browser_main_linux.cc	2023-05-06 18:59:33.922075000 +0300
@@ -57,11 +57,11 @@
 #if BUILDFLAG(IS_CHROMEOS)
   // No-op: Ash and Lacros Bluetooth DBusManager initialization depend on
   // FeatureList, and is done elsewhere.
+#elif defined(USE_DBUS)
+  bluez::BluezDBusManager::Initialize(nullptr /* system_bus */);
 #endif  // BUILDFLAG(IS_CHROMEOS)
 
 #if !BUILDFLAG(IS_CHROMEOS)
-  bluez::BluezDBusManager::Initialize(nullptr /* system_bus */);
-
   // Set up crypt config. This needs to be done before anything starts the
   // network service, as the raw encryption key needs to be shared with the
   // network service for encrypted cookie storage.
@@ -122,7 +122,7 @@
 void ChromeBrowserMainPartsLinux::PostDestroyThreads() {
 #if BUILDFLAG(IS_CHROMEOS)
   // No-op; per PostBrowserStart() comment, this is done elsewhere.
-#else
+#elif defined(USE_DBUS)
   bluez::BluezDBusManager::Shutdown();
   bluez::BluezDBusThreadManager::Shutdown();
 #endif  // BUILDFLAG(IS_CHROMEOS)
diff -uraN a/chrome/browser/metrics/bluetooth_metrics_provider.cc b/chrome/browser/metrics/bluetooth_metrics_provider.cc
--- a/chrome/browser/metrics/bluetooth_metrics_provider.cc	2023-04-27 00:42:53.546698600 +0300
+++ b/chrome/browser/metrics/bluetooth_metrics_provider.cc	2023-05-06 18:59:33.922075000 +0300
@@ -15,11 +15,15 @@
 BluetoothMetricsProvider::~BluetoothMetricsProvider() = default;
 
 bool BluetoothMetricsProvider::ProvideHistograms() {
+#ifdef USE_DBUS
   base::UmaHistogramEnumeration("Bluetooth.StackName",
                                 floss::features::IsFlossEnabled()
                                     ? BluetoothStackName::kFloss
                                     : BluetoothStackName::kBlueZ);
   return true;
+#else
+  return false;
+#endif
 }
 
 }  // namespace metrics
diff -uraN a/chrome/browser/platform_util_linux.cc b/chrome/browser/platform_util_linux.cc
--- a/chrome/browser/platform_util_linux.cc	2023-04-27 00:42:53.922731900 +0300
+++ b/chrome/browser/platform_util_linux.cc	2023-05-06 18:59:33.922075000 +0300
@@ -29,11 +29,15 @@
 // complain as Chromecast doesn't use (or depend on) //components/dbus.
 // TODO(crbug.com/1215474): Eliminate //chrome being visible in the GN structure
 // on Chromecast and remove the nogncheck below.
+#if defined(USE_DBUS)
 #include "components/dbus/thread_linux/dbus_thread_linux.h"  // nogncheck
+#endif
 #include "content/public/browser/browser_thread.h"
+#if defined(USE_DBUS)
 #include "dbus/bus.h"
 #include "dbus/message.h"
 #include "dbus/object_proxy.h"
+#endif
 #include "third_party/abseil-cpp/absl/types/optional.h"
 #include "url/gurl.h"
 
@@ -43,6 +47,7 @@
 
 namespace {
 
+#if defined(USE_DBUS)
 const char kMethodListActivatableNames[] = "ListActivatableNames";
 const char kMethodNameHasOwner[] = "NameHasOwner";
 
@@ -288,6 +293,7 @@
   base::CallbackListSubscription browser_shutdown_subscription_;
   base::WeakPtrFactory<ShowItemHelper> weak_ptr_factory_{this};
 };
+#endif
 
 void RunCommand(const std::string& command,
                 const base::FilePath& working_directory,
@@ -354,8 +360,10 @@
 }  // namespace internal
 
 void ShowItemInFolder(Profile* profile, const base::FilePath& full_path) {
+#if defined(USE_DBUS)
   DCHECK_CURRENTLY_ON(BrowserThread::UI);
   ShowItemHelper::GetInstance().ShowItemInFolder(profile, full_path);
+#endif
 }
 
 void OpenExternal(const GURL& url) {
diff -uraN a/device/bluetooth/adapter.cc b/device/bluetooth/adapter.cc
--- a/device/bluetooth/adapter.cc	2023-04-27 00:43:05.911790000 +0300
+++ b/device/bluetooth/adapter.cc	2023-05-06 18:59:33.923075000 +0300
@@ -109,7 +109,7 @@
   mojom::AdapterInfoPtr adapter_info = mojom::AdapterInfo::New();
   adapter_info->address = adapter_->GetAddress();
   adapter_info->name = adapter_->GetName();
-#if BUILDFLAG(IS_CHROMEOS) || BUILDFLAG(IS_LINUX)
+#if BUILDFLAG(IS_CHROMEOS) || (BUILDFLAG(IS_LINUX) && defined(USE_DBUS))
   adapter_info->system_name = adapter_->GetSystemName();
   adapter_info->floss = floss::features::IsFlossEnabled();
 #endif
diff -uraN a/device/bluetooth/bluetooth_adapter_factory.cc b/device/bluetooth/bluetooth_adapter_factory.cc
--- a/device/bluetooth/bluetooth_adapter_factory.cc	2023-04-27 00:43:05.915790300 +0300
+++ b/device/bluetooth/bluetooth_adapter_factory.cc	2023-05-06 19:00:48.276084194 +0300
@@ -42,7 +42,8 @@
   // instance even on platforms that would otherwise not support it.
   if (Get()->adapter_)
     return true;
-#if BUILDFLAG(IS_ANDROID) || BUILDFLAG(IS_WIN) || BUILDFLAG(IS_LINUX) || \
+#if BUILDFLAG(IS_ANDROID) || BUILDFLAG(IS_WIN) || \
+    (BUILDFLAG(IS_LINUX) && defined(USE_DBUS)) || \
     BUILDFLAG(IS_CHROMEOS) || BUILDFLAG(IS_APPLE)
   return true;
 #else
diff -uraN a/extensions/browser/api/bluetooth_low_energy/bluetooth_low_energy_api.cc b/extensions/browser/api/bluetooth_low_energy/bluetooth_low_energy_api.cc
--- a/extensions/browser/api/bluetooth_low_energy/bluetooth_low_energy_api.cc	2023-04-27 00:43:06.311825300 +0300
+++ b/extensions/browser/api/bluetooth_low_energy/bluetooth_low_energy_api.cc	2023-05-06 18:59:33.924075000 +0300
@@ -1319,7 +1319,7 @@
 // TODO: Ideally this should be handled by our feature system, so that this
 // code doesn't even compile on OSes it isn't being used on, but currently this
 // is not possible.
-#if !BUILDFLAG(IS_WIN)
+#if !BUILDFLAG(IS_WIN) && (!BUILDFLAG(IS_LINUX) || defined(USE_DBUS))
   base::WeakPtr<device::BluetoothLocalGattService> service =
       device::BluetoothLocalGattService::Create(
           event_router_->adapter(),
@@ -1348,6 +1348,7 @@
 }
 
 void BluetoothLowEnergyCreateCharacteristicFunction::DoWork() {
+#if !BUILDFLAG(IS_LINUX) || defined(USE_DBUS)
   device::BluetoothLocalGattService* service =
       event_router_->adapter()->GetGattService(params_->service_id);
   if (!service) {
@@ -1368,6 +1369,9 @@
 
   Respond(ArgumentList(apibtle::CreateCharacteristic::Results::Create(
       characteristic->GetIdentifier())));
+#else
+  Respond(Error(kErrorPlatformNotSupported));
+#endif
 }
 
 // createDescriptor:
@@ -1384,6 +1388,7 @@
 }
 
 void BluetoothLowEnergyCreateDescriptorFunction::DoWork() {
+#if !BUILDFLAG(IS_LINUX) || defined(USE_DBUS)
   device::BluetoothLocalGattCharacteristic* characteristic =
       event_router_->GetLocalCharacteristic(params_->characteristic_id);
   if (!characteristic) {
@@ -1399,6 +1404,9 @@
 
   Respond(ArgumentList(
       apibtle::CreateDescriptor::Results::Create(descriptor->GetIdentifier())));
+#else
+  Respond(Error(kErrorPlatformNotSupported));
+#endif
 }
 
 // registerService:
diff -uraN a/services/device/wake_lock/power_save_blocker/BUILD.gn b/services/device/wake_lock/power_save_blocker/BUILD.gn
--- a/services/device/wake_lock/power_save_blocker/BUILD.gn	2023-04-27 00:43:11.356270600 +0300
+++ b/services/device/wake_lock/power_save_blocker/BUILD.gn	2023-05-06 18:59:33.924075000 +0300
@@ -84,6 +84,9 @@
     visibility = [ ":power_save_blocker" ]
     sources = java_sources_needing_jni
   }
+  if (!is_chromeos && use_x11 && !use_dbus) {
+    sources += [ "power_save_blocker_ozone.cc" ]
+  }
 
   android_library("java") {
     # Conceptually, this should be visible only to //services/device:java.
diff -uraN a/ui/shell_dialogs/BUILD.gn b/ui/shell_dialogs/BUILD.gn
--- a/ui/shell_dialogs/BUILD.gn	2023-04-27 00:44:04.088924400 +0300
+++ b/ui/shell_dialogs/BUILD.gn	2023-05-06 18:59:33.925075001 +0300
@@ -55,7 +55,6 @@
       "shell_dialog_linux.h",
     ]
     deps += [
-      "//components/dbus/thread_linux",
       "//ui/linux:linux_ui",
     ]
     if (use_dbus) {
@@ -64,7 +63,10 @@
         "select_file_dialog_linux_portal.cc",
         "select_file_dialog_linux_portal.h",
       ]
-      deps += [ "//dbus" ]
+      deps += [
+        "//components/dbus/thread_linux",
+        "//dbus",
+      ]
     }
   }
 
