--- a/base/BUILD.gn	2023-03-07 10:20:53.334587000 +0300
+++ b/base/BUILD.gn	2023-03-10 18:14:56.172260401 +0300
@@ -1541,14 +1541,6 @@
     public_deps += [ ":nodebug_assertion" ]
   }
 
-  # Needed for <atomic> if using newer C++ library than sysroot, except if
-  # building inside the cros_sdk environment - use host_toolchain as a
-  # more robust check for this.
-  if (!use_sysroot && (is_android || is_chromeos || (is_linux && !is_castos)) &&
-      host_toolchain != "//build/toolchain/cros:host") {
-    libs += [ "atomic" ]
-  }
-
   if (use_allocator_shim) {
     sources += [
       "allocator/partition_allocator/shim/allocator_shim.cc",
--- a/build/config/linux/BUILD.gn	2023-03-07 10:20:53.750622000 +0300
+++ b/build/config/linux/BUILD.gn	2023-03-10 18:14:35.836260803 +0300
@@ -39,11 +39,6 @@
   if (is_chromeos) {
     defines = [ "OS_CHROMEOS" ]
   }
-
-  if ((!is_chromeos || default_toolchain != "//build/toolchain/cros:target") &&
-      (!use_custom_libcxx || current_cpu == "mipsel")) {
-    libs = [ "atomic" ]
-  }
 }
 
 config("libcap") {
