diff -uraN a/meson.build b/meson.build
--- a/meson.build	2019-12-18 21:36:00.000000000 +0300
+++ b/meson.build	2019-12-23 07:25:30.677890360 +0300
@@ -387,7 +387,8 @@
 endif
 
 # Android uses emutls for versions <= P/28. For USE_ELF_TLS we need ELF TLS.
-if host_machine.system() != 'windows' and (not with_platform_android or get_option('platform-sdk-version') >= 29)
+with_use_elf_tls = get_option('use-elf-tls')
+if with_use_elf_tls and not host_machine.system() != 'windows' or with_platform_android or get_option('platform-sdk-version') >= 29
   pre_args += '-DUSE_ELF_TLS'
 endif
 
diff -uraN a/meson_options.txt b/meson_options.txt
--- a/meson_options.txt	2019-12-18 21:36:00.000000000 +0300
+++ b/meson_options.txt	2019-12-23 07:25:49.426889370 +0300
@@ -366,3 +366,10 @@
   value : 25,
   description : 'Android Platform SDK version. Default: Nougat version.'
 )
+
+option(
+  'use-elf-tls',
+  type : 'boolean',
+  value : false,
+  description : 'Build support for initial-exec TLS model'
+)
