From 5398b5f3956e142fe8ccabc2580471a81cc0d7cf Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Mon, 4 Apr 2022 22:36:57 +0300
Subject: [PATCH 02/13] disable auto_mic by default. (mss)

---
 sound/pci/hda/hda_generic.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sound/pci/hda/hda_generic.c b/sound/pci/hda/hda_generic.c
index fc114e522..89633b2c7 100644
--- a/sound/pci/hda/hda_generic.c
+++ b/sound/pci/hda/hda_generic.c
@@ -141,7 +141,7 @@ static void parse_user_hints(struct hda_codec *codec)
 	val = snd_hda_get_bool_hint(codec, "auto_mute");
 	if (val >= 0)
 		spec->suppress_auto_mute = !val;
-	val = snd_hda_get_bool_hint(codec, "auto_mic");
+	val = 0;
 	if (val >= 0)
 		spec->suppress_auto_mic = !val;
 	val = snd_hda_get_bool_hint(codec, "line_in_auto_switch");
-- 
2.35.1

