From 00bd0cdda2c4e2e2a8427218b646ff9a9e17e0cf Mon Sep 17 00:00:00 2001
From: must_eat <mss@tutanota.de>
Date: Mon, 5 Dec 2022 21:46:00 +0300
Subject: [PATCH 2/9] hda: disable auto_mic by default.

---
 sound/pci/hda/hda_generic.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sound/pci/hda/hda_generic.c b/sound/pci/hda/hda_generic.c
index fc114e522..89633b2c7 100644
--- a/sound/pci/hda/hda_generic.c
+++ b/sound/pci/hda/hda_generic.c
@@ -141,7 +141,7 @@ static void parse_user_hints(struct hda_codec *codec)
 	val = snd_hda_get_bool_hint(codec, "auto_mute");
 	if (val >= 0)
 		spec->suppress_auto_mute = !val;
-	val = snd_hda_get_bool_hint(codec, "auto_mic");
+	val = 0;
 	if (val >= 0)
 		spec->suppress_auto_mic = !val;
 	val = snd_hda_get_bool_hint(codec, "line_in_auto_switch");
-- 
2.38.1

