From 68165d9eec4a6f400126f90bf1e551e68a63811b Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Mon, 4 Apr 2022 22:43:30 +0300
Subject: [PATCH 04/13] enable -O3 KBUILD_CFLAGS for all arches.
 (xanmod/xanmod)

---
 init/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index e9119bf54..98b648e3c 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1370,7 +1370,6 @@ config CC_OPTIMIZE_FOR_PERFORMANCE
 
 config CC_OPTIMIZE_FOR_PERFORMANCE_O3
 	bool "Optimize more for performance (-O3)"
-	depends on ARC
 	help
 	  Choosing this option will pass "-O3" to your compiler to optimize
 	  the kernel yet more for performance.
-- 
2.35.1

