From 7c01e4638bc758d97e655386e74626d1b5008c3e Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Tue, 11 Jan 2022 18:00:39 +0300
Subject: [PATCH 4/5] enable -O3.

---
 init/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 4b7bac10c..32f03ee98 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1370,7 +1370,6 @@ config CC_OPTIMIZE_FOR_PERFORMANCE
 
 config CC_OPTIMIZE_FOR_PERFORMANCE_O3
 	bool "Optimize more for performance (-O3)"
-	depends on ARC
 	help
 	  Choosing this option will pass "-O3" to your compiler to optimize
 	  the kernel yet more for performance.
-- 
2.34.1

