From 5d51ba4ccaf1ac70ee0b6dcc93645040dc75af96 Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Wed, 3 Aug 2022 21:54:41 +0300
Subject: [PATCH 04/12] enable -O3 KBUILD_CFLAGS for all arches.

---
 init/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index c7900e897..97c2df3ac 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1403,7 +1403,6 @@ config CC_OPTIMIZE_FOR_PERFORMANCE
 
 config CC_OPTIMIZE_FOR_PERFORMANCE_O3
 	bool "Optimize more for performance (-O3)"
-	depends on ARC
 	help
 	  Choosing this option will pass "-O3" to your compiler to optimize
 	  the kernel yet more for performance.
-- 
2.37.1

