From fbd25852dbb9be3c3d225bd97f68f16139b7f59d Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Wed, 25 May 2022 18:19:07 +0300
Subject: [PATCH 04/12] enable -O3 KBUILD_CFLAGS for all arches.
 (xanmod5.17/xanmod)

---
 init/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index ddcbefe53..ddd31154b 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1374,7 +1374,6 @@ config CC_OPTIMIZE_FOR_PERFORMANCE
 
 config CC_OPTIMIZE_FOR_PERFORMANCE_O3
 	bool "Optimize more for performance (-O3)"
-	depends on ARC
 	help
 	  Choosing this option will pass "-O3" to your compiler to optimize
 	  the kernel yet more for performance.
-- 
2.36.0

