From 1ee54d3bed56e626d98b2ba997717888ac398870 Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Tue, 16 Nov 2021 15:42:16 +0300
Subject: [PATCH 06/08] enable O3.

---
 init/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 11f8a845f..806c196d5 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1365,7 +1365,6 @@ config CC_OPTIMIZE_FOR_PERFORMANCE
 
 config CC_OPTIMIZE_FOR_PERFORMANCE_O3
 	bool "Optimize more for performance (-O3)"
-	depends on ARC
 	help
 	  Choosing this option will pass "-O3" to your compiler to optimize
 	  the kernel yet more for performance.
-- 
2.31.1

