From 22609e6f70d25a52af0bd052f43366b4b19a74cd Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Mon, 4 Apr 2022 22:48:47 +0300
Subject: [PATCH 09/13] remove MAXSMP check for for mglru. (mss)

---
 mm/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/Kconfig b/mm/Kconfig
index 7fd84e038..10e078791 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -897,7 +897,7 @@ config LRU_GEN
 	bool "Multi-Gen LRU"
 	depends on MMU
 	# the following options can use up the spare bits in page flags
-	depends on !MAXSMP && (64BIT || !SPARSEMEM || SPARSEMEM_VMEMMAP)
+	depends on 64BIT || !SPARSEMEM || SPARSEMEM_VMEMMAP
 	help
 	  A high performance LRU implementation for memory overcommit. See
 	  Documentation/admin-guide/mm/multigen_lru.rst for details.
-- 
2.35.1

