From c8d3be711360e0249a8e6a0f13b201ca05a5fbfa Mon Sep 17 00:00:00 2001
From: mssx86 <mss@tutanota.de>
Date: Wed, 8 Sep 2021 08:33:08 +0300
Subject: [PATCH 10/11] set rq_affinity to force full multithreading I/O
 requests.

---
 include/linux/blkdev.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/linux/blkdev.h b/include/linux/blkdev.h
index d3afea47a..29334d6e7 100644
--- a/include/linux/blkdev.h
+++ b/include/linux/blkdev.h
@@ -606,6 +606,7 @@ struct request_queue {
 
 #define QUEUE_FLAG_MQ_DEFAULT	((1 << QUEUE_FLAG_IO_STAT) |		\
 				 (1 << QUEUE_FLAG_SAME_COMP) |		\
+				 (1 << QUEUE_FLAG_SAME_FORCE) |		\
 				 (1 << QUEUE_FLAG_NOWAIT))
 
 void blk_queue_flag_set(unsigned int flag, struct request_queue *q);
-- 
2.33.0

