diff -uraN a/sound/pci/hda/hda_generic.c b/sound/pci/hda/hda_generic.c
--- a/sound/pci/hda/hda_generic.c	2021-09-18 14:44:05.000000000 +0300
+++ b/sound/pci/hda/hda_generic.c	2021-09-19 12:51:19.228831250 +0300
@@ -135,7 +135,7 @@
 	val = snd_hda_get_bool_hint(codec, "auto_mute");
 	if (val >= 0)
 		spec->suppress_auto_mute = !val;
-	val = snd_hda_get_bool_hint(codec, "auto_mic");
+	val = 0;
 	if (val >= 0)
 		spec->suppress_auto_mic = !val;
 	val = snd_hda_get_bool_hint(codec, "line_in_auto_switch");
