diff -uraN a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2022-10-12 05:08:13.000000000 +0300
+++ b/CMakeLists.txt	2023-04-20 18:35:11.755235358 +0300
@@ -159,9 +159,7 @@
 
 find_package(PkgConfig)
 # Check for libraries required for building
-find_library(LIBC NAMES libc.so.6 REQUIRED)
 find_library(NUMA NAMES numa REQUIRED)
-message(STATUS "LIBC:" ${LIBC})
 message(STATUS "NUMA:" ${NUMA})
 
 ## If environment variable DRM_DIR is set, the script
@@ -183,7 +181,7 @@
 include_directories(${DRM_INCLUDE_DIRS})
 
 target_link_libraries ( ${HSAKMT_TARGET}
-  PRIVATE ${DRM_LDFLAGS} ${DRM_AMDGPU_LDFLAGS} pthread rt ${LIBC} numa
+  PRIVATE ${DRM_LDFLAGS} ${DRM_AMDGPU_LDFLAGS} pthread rt numa
 )
 
 target_compile_options(${HSAKMT_TARGET} PRIVATE ${DRM_CFLAGS} ${HSAKMT_C_FLAGS})
diff -uraN a/src/globals.c b/src/globals.c
--- a/src/globals.c	2022-10-12 05:08:13.000000000 +0300
+++ b/src/globals.c	2023-04-20 18:35:26.319234379 +0300
@@ -32,5 +32,7 @@
 unsigned long system_properties_count;
 pthread_mutex_t hsakmt_mutex = PTHREAD_MUTEX_INITIALIZER;
 bool is_dgpu;
+#if !defined(PAGE_SIZE)
 int PAGE_SIZE;
+#endif
 int PAGE_SHIFT;
diff -uraN a/src/libhsakmt.h b/src/libhsakmt.h
--- a/src/libhsakmt.h	2022-10-12 05:08:13.000000000 +0300
+++ b/src/libhsakmt.h	2023-04-20 18:35:26.319234379 +0300
@@ -56,7 +56,9 @@
 	do { if ((minor) > kfd_version_info.KernelInterfaceMinorVersion)\
 		return HSAKMT_STATUS_NOT_SUPPORTED; } while (0)
 
+#if !defined(PAGE_SIZE)
 extern int PAGE_SIZE;
+#endif
 extern int PAGE_SHIFT;
 
 /* VI HW bug requires this virtual address alignment */
diff -uraN a/src/openclose.c b/src/openclose.c
--- a/src/openclose.c	2022-10-12 05:08:13.000000000 +0300
+++ b/src/openclose.c	2023-04-20 18:35:26.319234379 +0300
@@ -108,7 +108,9 @@
 
 static inline void init_page_size(void)
 {
+#if !defined(PAGE_SIZE)
 	PAGE_SIZE = sysconf(_SC_PAGESIZE);
+#endif
 	PAGE_SHIFT = ffs(PAGE_SIZE) - 1;
 }
 
diff -uraN a/src/topology.c b/src/topology.c
--- a/src/topology.c	2022-10-12 05:08:13.000000000 +0300
+++ b/src/topology.c	2023-04-20 18:39:31.150217917 +0300
@@ -42,6 +42,8 @@
 #include "libhsakmt.h"
 #include "fmm.h"
 
+#include <sys/param.h>
+
 /* Number of memory banks added by thunk on top of topology
  * This only includes static heaps like LDS, scratch and SVM,
  * not for MMIO_REMAP heap. MMIO_REMAP memory bank is reported
