diff -uraN a/os/os.hpp b/os/os.hpp
--- a/os/os.hpp	2023-01-16 19:23:20.000000000 +0300
+++ b/os/os.hpp	2023-04-21 15:08:30.919244999 +0300
@@ -29,12 +29,17 @@
 
 #if defined(__linux__)
 #include <sched.h>
+#include <libgen.h>
 #endif
 
 #ifdef _WIN32
 #include <Basetsd.h>  // For KAFFINITY
 #endif                // _WIN32
 
+#ifndef __cpu_mask
+typedef unsigned long __cpu_mask;
+#endif
+
 // Smallest supported VM page size.
 #define MIN_PAGE_SHIFT 12
 #define MIN_PAGE_SIZE (1UL << MIN_PAGE_SHIFT)
diff -uraN a/os/os_posix.cpp b/os/os_posix.cpp
--- a/os/os_posix.cpp	2023-01-16 19:23:20.000000000 +0300
+++ b/os/os_posix.cpp	2023-04-21 15:08:30.919244999 +0300
@@ -394,9 +394,11 @@
       for (int i = 0; i < processorCount_; i++) {
         CPU_SET(i, &cpuset);
       }
+#ifdef pthread_attr_setaffinity_np
       if (0 != pthread_attr_setaffinity_np(&threadAttr, sizeof(cpu_set_t), &cpuset)) {
         fatal("pthread_attr_setaffinity_np failed to set affinity");
       }
+#endif
     }
   }
 
diff -uraN a/platform/activity.cpp b/platform/activity.cpp
--- a/platform/activity.cpp	2023-01-16 19:23:20.000000000 +0300
+++ b/platform/activity.cpp	2023-04-21 15:08:33.479244827 +0300
@@ -32,7 +32,7 @@
 #if USE_PROF_API
 
 #if defined(__linux__)
-__thread activity_correlation_id_t correlation_id __attribute__((tls_model("initial-exec"))) = 0;
+__thread activity_correlation_id_t correlation_id = 0;
 #elif defined(_WIN32)
 __declspec(thread) activity_correlation_id_t correlation_id = 0;
 #endif  // defined(_WIN32)
diff -uraN a/platform/activity.hpp b/platform/activity.hpp
--- a/platform/activity.hpp	2023-01-16 19:23:20.000000000 +0300
+++ b/platform/activity.hpp	2023-04-21 15:08:33.479244827 +0300
@@ -46,7 +46,7 @@
     report_activity;
 
 #if defined(__linux__)
-extern __thread activity_correlation_id_t correlation_id __attribute__((tls_model("initial-exec")));
+extern __thread activity_correlation_id_t correlation_id;
 #elif defined(_WIN32)
 extern __declspec(thread) activity_correlation_id_t correlation_id;
 #endif  // defined(_WIN32)
diff -uraN a/thread/thread.cpp b/thread/thread.cpp
--- a/thread/thread.cpp	2023-01-16 19:23:20.000000000 +0300
+++ b/thread/thread.cpp	2023-04-21 15:08:33.479244827 +0300
@@ -121,7 +121,7 @@
 
 namespace details {
 
-__thread Thread* thread_ __attribute__((tls_model("initial-exec")));
+__thread Thread* thread_;
 
 }  // namespace details
 
diff -uraN a/thread/thread.hpp b/thread/thread.hpp
--- a/thread/thread.hpp	2023-01-16 19:23:20.000000000 +0300
+++ b/thread/thread.hpp	2023-04-21 15:08:33.479244827 +0300
@@ -186,7 +186,7 @@
 
 #if defined(__linux__)
 
-extern __thread Thread* thread_ __attribute__((tls_model("initial-exec")));
+extern __thread Thread* thread_;
 
 static inline Thread* currentThread() { return thread_; }
 
