diff -uraN a/khronos/icd/loader/icd_platform.h b/khronos/icd/loader/icd_platform.h
--- a/khronos/icd/loader/icd_platform.h	2022-10-20 04:40:57.000000000 +0300
+++ b/khronos/icd/loader/icd_platform.h	2023-04-21 15:11:43.883232025 +0300
@@ -26,7 +26,7 @@
 #ifdef __ANDROID__
 #define ICD_VENDOR_PATH "/system/vendor/Khronos/OpenCL/vendors/";
 #else
-#define ICD_VENDOR_PATH "/etc/OpenCL/vendors/";
+#define ICD_VENDOR_PATH "/opt/rocm/vendors/";
 #endif // ANDROID
 
 #elif defined(_WIN32)
