diff --git a/src/core/CdStreamPosix.cpp b/src/core/CdStreamPosix.cpp
index 30fe06a0..bc9129eb 100644
--- a/src/core/CdStreamPosix.cpp
+++ b/src/core/CdStreamPosix.cpp
@@ -13,6 +13,7 @@
 #include <fcntl.h>
 #include <sys/resource.h>
 #include <stdarg.h>
+#include <limits.h>
 
 #ifdef __linux__
 #include <sys/syscall.h>
diff --git a/src/skel/crossplatform.h b/src/skel/crossplatform.h
index 9b43bcaf..0413e4d4 100644
--- a/src/skel/crossplatform.h
+++ b/src/skel/crossplatform.h
@@ -1,4 +1,5 @@
 #include <time.h>
+#include <limits.h>
 
 // This is the common include for platform/renderer specific skeletons(glfw.cpp, win.cpp etc.) and using cross platform things (like Windows directories wrapper, platform specific global arrays etc.) 
 // Functions that's different on glfw and win but have same signature, should be located on platform.h.
